### PR TITLE
Added error block and faulty node caching functions to improve stability, and increased the configuration of the p2p node selection strategy when synchronizing blocks

### PR DESCRIPTION
**Pull request by wangchao0222**
## Description

1. 同步最长链时增加了错误区块缓存功能
2. 增加同步区块时过滤模式的配置和过滤factor配置，通过配置可以选择BucketsFilterWithFactor模式或者NearestBucketFilter模式
3. 重现编写BucketsFilterWithFactor的Filter函数，优化原始的过滤的逻辑，优化步骤

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Brief of your solution

Please describe your solution to solve the issue or feature request.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

### CODE REVIEW

File: kernel/engines/xuperos/miner/sync.go [hunk 1]
```diff
@@ -103,11 +105,51 @@ func (t *Miner) getMaxBlockHeight(ctx xctx.XContext) (string, int64, error) {
...
+					t.faultBlockIdCache.Set(string(status.LedgerMeta.TipBlockid), response.Header.From, faultBlockIdCacheExpired)
+				}
+				// 增加peerId记录错误次数
+				count, errInc := t.faultPeerIdCache.IncrementInt32(response.Header.From, int32(1))
+				if errInc != nil {
+					count = 1
+					t.faultPeerIdCache.Set(response.Header.From, count, faultPeerIdCacheExpired)
+				}
+				ctx.GetLog().Debug("faultPeerIdCache Increment count", "curPeerIdStr", curPeerIdStr, "count", count, "errInc", errInc)
+				continue
```
Comment (chenfengjin) [inline kernel/engines/xuperos/miner/sync.go:127]: FYI: 
if 子句结尾有 continue 的话，可以不用 else 子句，减少if 的嵌套层次

File: kernel/engines/xuperos/miner/sync.go [hunk 2]
```diff
@@ -103,11 +105,51 @@ func (t *Miner) getMaxBlockHeight(ctx xctx.XContext) (string, int64, error) {
...
+			} else {
+				// 检查peerId是否超过故障次数
+				countInterface, hasPeer := t.faultPeerIdCache.Get(response.Header.From)
+				if hasPeer {
```
Comment (chenfengjin) [inline kernel/engines/xuperos/miner/sync.go:131]: FYI: 
if ! hasPeer {
continue 
} 可以 if 的减少嵌套层次

File: kernel/engines/xuperos/miner/sync.go [hunk 3]
```diff
@@ -103,11 +105,51 @@ func (t *Miner) getMaxBlockHeight(ctx xctx.XContext) (string, int64, error) {
...
+					count, okConvert := countInterface.(int32)
+					if !okConvert {
```
Comment (chenfengjin) [inline kernel/engines/xuperos/miner/sync.go:133]: 如果不是int32，这里还需要继续处理吗？

**Review by chenfengjin — APPROVED** [review]

File: kernel/engines/xuperos/miner/sync.go (hunk 3 above)
Comment (wangchao0222) [reply kernel/engines/xuperos/miner/sync.go:133]: 这里countInterface一定是int32类型

**Review by godeamon — APPROVED** [review]

**Pull request merged**